File: api/v1/schemas/volume_breakout.py
# -*- coding: utf-8 -*-
"""放量检测 API 响应模型"""

from typing import List
from pydantic import BaseModel


class BreakoutStock(BaseModel):
    code: str
    name: str
    volume: float
    yesterday_volume: float
    avg_3d_volume: float = 0
    ratio_vs_yesterday: float
    ratio_vs_3d_avg: float = 0
    sector_name: str = ""
    concept_names: str = ""


class BreakoutETF(BaseModel):
    code: str
    name: str
    volume: float
    yesterday_volume: float
    ratio_vs_yesterday: float


class BreakoutSector(BaseModel):
    name: str
    agg_volume: float
    yesterday_agg_volume: float
    ratio: float
    constituent_count: int


class BreakoutConcept(BaseModel):
    code: str = ""
    name: str
    agg_volume: float
    yesterday_agg_volume: float
    ratio: float
    constituent_count: int


class WordCloudWord(BaseModel):
    text: str
    value: int


class BreakoutResponse(BaseModel):
    date: str
    stock_count: int
    etf_count: int
    sector_count: int
    concept_count: int
    stocks: List[BreakoutStock]
    etfs: List[BreakoutETF]
    sectors: List[BreakoutSector]
    concepts: List[BreakoutConcept]
    sector_words: List[WordCloudWord]
    concept_words: List[WordCloudWord]


class BreakoutDatesResponse(BaseModel):
    dates: List[str]

File: apps/dsa-web/src/api/volumeBreakout.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export interface WordCloudWord {
  text: string;
  value: number;
}

export interface BreakoutStock {
  code: string;
  name: string;
  volume: number;
  yesterdayVolume: number;
  avg3dVolume: number;
  ratioVsYesterday: number;
  ratioVs3dAvg: number;
  sectorName: string;
  conceptNames: string;
}

export interface BreakoutETF {
  code: string;
  name: string;
  volume: number;
  yesterdayVolume: number;
  ratioVsYesterday: number;
}

export interface BreakoutSector {
  name: string;
  aggVolume: number;
  yesterdayAggVolume: number;
  ratio: number;
  constituentCount: number;
}

export interface BreakoutConcept {
  code: string;
  name: string;
  aggVolume: number;
  yesterdayAggVolume: number;
  ratio: number;
  constituentCount: number;
}

export interface BreakoutResponse {
  date: string;
  stockCount: number;
  etfCount: number;
  sectorCount: number;
  conceptCount: number;
  stocks: BreakoutStock[];
  etfs: BreakoutETF[];
  sectors: BreakoutSector[];
  concepts: BreakoutConcept[];
  sectorWords: WordCloudWord[];
  conceptWords: WordCloudWord[];
}

export interface BreakoutDatesResponse {
  dates: string[];
}

export const volumeBreakoutApi = {
  getResults: async (date?: string): Promise<BreakoutResponse> => {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/volume-breakout/results', {
      params: { date },
    });
    return toCamelCase<BreakoutResponse>(response.data);
  },

  getDates: async (days = 30): Promise<BreakoutDatesResponse> => {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/volume-breakout/dates', {
      params: { days },
    });
    return toCamelCase<BreakoutDatesResponse>(response.data);
  },
};
